test/opt/assembly_builder.h

```diff
@@ -0,0 +1,207 @@
+// to this code builder will be assigned with a global name through OpName
+// instruction. The name is extracted from the definition instruction.
+//  E.g. adding instruction: %var_a = OpConstant %int 2, will also add an
+//  instruction: OpName %var_a, "var_a".
+//
+// Note that the name must not be used on more than defined IDs and friendly-name
+// disassembling must be enabled so that OpName instructions will be respected.
+class AssemblyBuilder {
+ public:
+  // Initalize an minimal SPIR-V assembly code as the template. The minimal
+  // contains an empty main function and some predefined names for the main
+  // function.
+  AssemblyBuilder()
+      : global_preamble_vec_({
+            // clang-format off
+                  "OpCapability Shader",
+                  "OpCapability Float64",
+             "%1 = OpExtInstImport \"GLSL.std.450\"",
+                  "OpMemoryModel Logical GLSL450",
+                  "OpEntryPoint Vertex %main \"main\"",
+            // clang-format on
+        }),
+        names_vec_({
+            // clang-format off
+                  "OpName %main \"main\"",
+                  "OpName %main_func_entry_block \"main_func_entry_block\"",
+                  "OpName %main_func_type \"main_func_type\"",
+            // clang-format on
+        }),
+        types_consts_globals_vec_({
+            // clang-format off
+          "%void = OpTypeVoid",
+"%main_func_type = OpTypeFunction %void",
+            // clang-format on
+        }),
+        main_func_preamble_vec_({
+            // clang-format off
+          "%main = OpFunction %void None %main_func_type",
+"%main_func_entry_block = OpLabel"
+            // clang-format on
+        }),
+        main_func_entry_block_vec_(),
+        main_func_postamble_vec_({
+            // clang-format off
+                  "OpReturn",
+                  "OpFunctionEnd",
+            // clang-format on
+        }),
+        names_({{"main", 0}, {"main_func_entry_block", 1}, {"main_func_type", 2}}) {}
+
+  // Appends instructions to the types-constants-globals section.
+  AssemblyBuilder& AppendTypesConstantsGlobals(
+      const std::vector<std::string>& vec_asm_code) {
+    AddNamesFor(vec_asm_code);
+    types_consts_globals_vec_.insert(types_consts_globals_vec_.end(),
+                                     vec_asm_code.begin(), vec_asm_code.end());
+    return *this;
+  }
+
+  // Appends instructions to the main function block, start within
+  // "main_func_entry_block".
+  AssemblyBuilder& AppendInMain(const std::vector<std::string>& vec_asm_code) {
+    AddNamesFor(vec_asm_code);
+    main_func_entry_block_vec_.insert(main_func_entry_block_vec_.end(),
+                                      vec_asm_code.begin(), vec_asm_code.end());
+    return *this;
+  }
+
+  // Get the SPIR-V assembly code in string.
+  std::string GetCode() const {
+    std::stringstream ss;
+    for (auto& line : global_preamble_vec_) {
```

Comment:
This is a lot of repeated code.
I wonder if instead we should have had an array of vectors, and the first level is indexed by segment enum like kGlobalPreamble, kNames, kTypesConstantsGLobals, kMainFuncPreamble etc.

But not worth changing now.  We could think about this later.